### PR TITLE
Update the validation test of Assignment, and update tests on group as…

### PR DESCRIPTION
**Pull request by YolandaZIK**
-line 696:
`clone` should be able to cloned correctly and `clone.grouping.size` should be equal to `assignment.grouping.size`

-line 674:
Assignment should return the correct group of a given student belongs to

`@assignment.group_by(@membership.user))` returns `nil`
Which means that `@assignment` cannot return the proper grouping 

Therefore, it cannot return the group which is expected to be the same as `@membership.grouping.group`

### CODE REVIEW

File: spec/models/assignment_spec.rb [hunk 1]
```diff
@@ -563,6 +636,71 @@
     end
   end
 
+context 'An Assignment' do
+  before :each do
+    @assignment = create(:assignment, group_name_autogenerated: false)
```
Comment (david-yz-liu) [inline spec/models/assignment_spec.rb]: This needs to be a group assignment for the test at line 674 to pass. You can make this change by passing `group_max: 2` when creating the assignment.

File: spec/models/assignment_spec.rb [hunk 2]
```diff
@@ -563,6 +636,71 @@
...
+    end
+
+    it "be able to have it's groupings cloned correctly" do
+      clone = create(:assignment, {group_min: 1, group_max: 1})
+      number = StudentMembership.all.size + TaMembership.all.size
+      clone.clone_groupings_from(@assignment.id)
+
+      expect(@assignment.group_min).to eql(clone.group_min)
+      expect(@assignment.group_max).to eql(clone.group_max)
+      expect(@assignment.groupings.size).to eql(clone.groupings.size)
```
Comment (david-yz-liu) [inline spec/models/assignment_spec.rb]: The logic here is actually good. The problem is that `clone.groupings` needs to be "reloaded" to obtain the updated value (with the 5 groups created). You can do this by calling `clone.groupings.reload` above this line.

File: spec/models/assignment_spec.rb [hunk 3]
```diff
@@ -61,28 +62,35 @@
...
+      end
     end
-    it 'should not be a peer review if there is no parent_assignment_id' do
-      assignment = create(:assignment)
-      expect(assignment.parent_assignment_id).to be_nil
-      expect(assignment.is_peer_review?).to be false
+
+    it 'should catch an invalid date' do
+      assignment = create(:assignment, due_date: '2020/02/31')  #31st day of february
+      !expect(assignment).to be_valid
```
Comment (mishaschwartz) [inline spec/models/assignment_spec.rb]: You should use the `not_to` method to get the inverse expectation for your test:

```
expect(assignment).not_to be_valid
```

see [here](http://rspec.info/documentation/3.6/rspec-expectations/RSpec/Matchers.html) for more examples

File: spec/models/assignment_spec.rb [hunk 4]
```diff
@@ -545,6 +562,62 @@
...
+
+      context 'another' do
+        before(:each) do
+          @section = create(:section, name: "section_name2")
+          create(:section_due_date, section: @section, assignment: @assignment,
+                 due_date: 1.day.ago)
+          student = create(:Student, section: @section)
+          @grouping = create(:grouping, assignment: @assignment)
+          create(:StudentMembership, grouping: @grouping, user: student,
+                 membership_status: StudentMembership::STATUSES[:inviter])
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: Put `grouping: ...` on a new line aligned with `:StudentMembership` to fix this comment.

File: spec/models/assignment_spec.rb [hunk 5]
```diff
@@ -563,6 +641,172 @@
...
+                 membership_status: StudentMembership::STATUSES[:accepted])
+        end
+      end
+
+      it "be able to have it's groupings cloned correctly" do
+        clone = create(:assignment, {group_min: 1, group_max: 1})
+        number = StudentMembership.all.size + TaMembership.all.size
+        clone.clone_groupings_from(@assignment.id)
+
+        clone.groupings.reload  # clone.groupings needs to be "reloaded" to obtain the updated value (5 groups created)
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: I think for this one you need to delete one space before the `#`.

File: spec/models/assignment_spec.rb [hunk 6]
```diff
@@ -1183,8 +1499,8 @@
             fields.push(result.total_mark)
             @assignment.get_criteria(:all, :rubric).each do |criterion|
               mark = result.marks
-                .find_by_markable_id_and_markable_type(criterion.id,
-                                                       'RubricCriterion')
+                       .find_by_markable_id_and_markable_type(criterion.id,
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: This means you need to align the `.` on this line with the `.` in `result.marks` on the previous line.

File: spec/models/assignment_spec.rb [hunk 7]
```diff
@@ -563,6 +606,279 @@
...
+            end
+          end
+
+          it 'be able to get_repo_checkout_commands' do
+            submissions = grouping_count(@assignment.groupings) # filter out without submission
+            expect(submissions).to eql @assignment.get_repo_checkout_commands.size
+          end
+        end
+      end
+  end
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: It looks like this line needs to be indented by 2 more spaces.

File: spec/models/assignment_spec.rb [hunk 8]
```diff
@@ -563,6 +606,279 @@
...
+
+        context 'with two groups of a single student each' do
+          before :each do
+            2.times do
+              g = create(:grouping, assignment: @assignment)
+              # StudentMembership.make({grouping: g,membership_status: StudentMembership::STATUSES[:inviter] } )
+              s = create(:submission, grouping: g)
+              r = s.get_latest_result
+              2.times do
+                create(:rubric_mark, result: r)  # this is create marks under rubric criterion
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: Similar to above, delete one space before `#`

File: spec/models/assignment_spec.rb [hunk 9]
```diff
@@ -563,6 +606,279 @@
...
+                                      grouping: @grouping,
+                                      membership_status: StudentMembership::STATUSES[:inviter])
+          @submission = create(:submission, grouping: @grouping)
+        end
+
+        it 'be able to get a list of repository access URLs for each group' do
+          expected_string = ''
+          @assignment.groupings.each do |grouping|
+            group = grouping.group
+            expected_string += [group.group_name,group.repository_external_access_url].to_csv
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: This one should be pretty easy to fix, just add spaces after each `,`

File: spec/models/assignment_spec.rb [hunk 10]
```diff
@@ -61,28 +63,38 @@
...
-      assignment = create(:assignment)
-      expect(assignment.parent_assignment_id).to be_nil
-      expect(assignment.is_peer_review?).to be false
+
+    it 'should catch an invalid date' do
+      assignment = Assignment.new(due_date: '2020/02/31') #31st day of february
+      expect(assignment).not_to be_valid
     end
+
+
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: (Delete this line)

File: spec/models/assignment_spec.rb [hunk 11]
```diff
@@ -61,28 +63,38 @@
...
+        expect(@assignment.is_peer_review?).to be false
+      end
     end
-    it 'should not be a peer review if there is no parent_assignment_id' do
-      assignment = create(:assignment)
-      expect(assignment.parent_assignment_id).to be_nil
-      expect(assignment.is_peer_review?).to be false
+
+    it 'should catch an invalid date' do
+      assignment = Assignment.new(due_date: '2020/02/31') #31st day of february
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: This means you need to add a space after the `#`

File: spec/models/assignment_spec.rb [hunk 12]
```diff
@@ -563,6 +606,279 @@
...
+                             group_max: 1,
+                             student_form_groups: false,
+                             invalid_override: true,
+                             due_date: 2.days.ago,
+                             created_at: 42.days.ago)
+      end
+
+      def grouping_count(groupings)
+        submissions = 0
+        for grouping in groupings do
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: This means do `groupings.each do |grouping|`

File: spec/models/assignment_spec.rb [hunk 13]
```diff
@@ -563,6 +606,279 @@
...
+        before :each do
+          @target = create(:assignment, group_min: 1, group_max: 1)
+        end
+
+        it 'clone all three members if none are hidden' do
+          @target.clone_groupings_from(@source.id)
+          3.times do |index|
+            expect @members[index].user.has_accepted_grouping_for?(@target.id)
+          end
+          @group.groupings.reload  #(is it necessary to call reload here?)
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: Remove the comment altogether.

File: spec/models/assignment_spec.rb [hunk 14]
```diff
@@ -563,6 +606,279 @@
...
+      end
+    end
+    context 'with a group with 3 accepted students' do
+      before :each do
+        @grouping = create(:grouping, assignment: @assignment)
+        @members = []
+        3.times do
+          @members.push create(:student_membership,
+                               membership_status: StudentMembership::STATUSES[:accepted],
+                               grouping: @grouping )
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: Delete the space before the `)`

File: spec/models/assignment_spec.rb [hunk 15]
```diff
@@ -1,4 +1,5 @@
 describe Assignment do
+
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: (Delete this line)

File: spec/models/assignment_spec.rb [hunk 16]
```diff
@@ -1183,8 +1451,8 @@
             fields.push(result.total_mark)
             @assignment.get_criteria(:all, :rubric).each do |criterion|
               mark = result.marks
-                .find_by_markable_id_and_markable_type(criterion.id,
-                                                       'RubricCriterion')
+                       .find_by_markable_id_and_markable_type(criterion.id,
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: (Duplicated from above)

File: spec/models/assignment_spec.rb [hunk 17]
```diff
@@ -563,6 +585,273 @@
...
+            end
+          end
+
+          it 'be able to get_repo_checkout_commands' do
+            submissions = grouping_count(@assignment.groupings) # filter out without submission
+            expect(submissions).to eql @assignment.get_repo_checkout_commands.size
+          end
+        end
+      end
+  end
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: (Duplicated from above)

File: spec/models/assignment_spec.rb [hunk 18]
```diff
@@ -545,6 +510,63 @@
...
+
+    context 'with a section' do
+      before(:each) do
+        @assignment = create(:assignment, due_date: 2.days.ago, section_due_dates_type: true)
+        @section = create(:section, name: 'section_name')
+        create(:section_due_date, section: @section, assignment: @assignment, due_date: 1.day.ago)
+        student = create(:student, section: @section)
+        @grouping = create(:grouping, assignment: @assignment)
+        create(:student_membership,
+                grouping: @grouping,
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/models/assignment_spec.rb]: Close, but need to delete one space before `grouping` so that it's aligned with the `:`

**Pull request merged**